tangerine/.dir-locals.el

```diff
@@ -0,0 +1,4 @@
+((auto-mode-alist . (("\\.h\\'" . c++-mode)))
+ (c++-mode . ((c-file-style . "linux")
```

Comment:
That helps! I'm not intimately familiar with the details of the various `c-file-style`s, but the only thing I noticed is that it seems `c-basic-offset` should be `4` rather than `8`. I'll try it and see if it works better.

Reply:
I had a look through what options emacs provides for `c-file-style`, and none of them really match the style guide I'm using.  Do you know if the formatting parameters can be set individually?

I think the main points are:
- Indent with tabs, one tab per indentation level.
- Braces are always on their own line at the same indentation level as the surrounding scope.
- `TitleCase` is used for all class names, function names, variable names, constants, and so on.
- Variable names should be descriptive and useful.
- Single digit variables may be lower case for loop counters, but should generally be uppercase otherwise.
- Single digit variable names should be avoided in favor of more descriptive names, but they do show up as loop counters, coordinates, and sometimes math terms.
- `UPPER_SNAKE_CASE` is used for preprocessor definitions.
- `snake_case` is used for labels.
- `goto` is acceptable when it improves readability.
- There's no line limit, but the code should probably read ok on a modest 1080p monitor in a full screen window with a reasonable font size.

I think `.dir-locals.el` should be made as accurate as possible otherwise it would be inappropriate to check into the repository.

Reply:
Do you actually "set your tab size to 4 characters" as the [Unreal guide](https://docs.unrealengine.com/5.1/en-US/epic-cplusplus-coding-standard-for-unreal-engine/#tabsandindenting) says? I tried changing `c-basic-offset` to 4, but it seems wrong, and likewise GitHub displays, say, these lines: https://github.com/Aeva/tangerine/blob/24cfebffdc499982ae383aabbeb253d17d5e2c29/tangerine/colors.cpp#L185-L187 with `const` aligned with the `o` of `StatusCode`, so 8 characters.

I think the Emacs C style is really limited to the indentation aspects, mostly. (I should say that, while I use Emacs, I'm by no means an Emacs wizard, and I also write C family languages as rarely as possible.) I haven't seen anything that the `"linux"` style indents incorrectly, but maybe I don't know the subtleties to look for.

I asked Emacs to guess the style in `gl_boilerplate.cpp`, and this is what it produced. AIUI, the lines labeled with `; Guessed value` are the changes from the base `"linux"` style:
```elisp
(c-add-style "STYLE NAME HERE"
             '("linux"
               (c-basic-offset . 8)     ; Guessed value
               (c-offsets-alist
                (block-close . 0)       ; Guessed value
                (brace-list-close . 0)  ; Guessed value
                (brace-list-entry . 0)  ; Guessed value
                (brace-list-intro . +)  ; Guessed value
                (case-label . 0)        ; Guessed value
                (defun-block-intro . +) ; Guessed value
                (defun-close . 0)       ; Guessed value
                (defun-open . 0)        ; Guessed value
                (else-clause . 0)       ; Guessed value
                (member-init-cont . -2) ; Guessed value
                (member-init-intro . +) ; Guessed value
                (statement . 0)         ; Guessed value
                (statement-block-intro . +) ; Guessed value
                (statement-case-intro . +) ; Guessed value
                (statement-cont . +)       ; Guessed value
                (substatement-open . 0)    ; Guessed value
                (topmost-intro . 0)        ; Guessed value
                (access-label . -)
                (annotation-top-cont . 0)
                (annotation-var-cont . +)
                (arglist-close . c-lineup-close-paren)
                (arglist-cont c-lineup-gcc-asm-reg 0)
                (arglist-cont-nonempty . c-lineup-arglist)
                (arglist-intro . +)
                (block-open . 0)
                (brace-entry-open . 0)
                (brace-list-open . 0)
                (c . c-lineup-C-comments)
                (catch-clause . 0)
                (class-close . 0)
                (class-open . 0)
                (comment-intro . c-lineup-comment)
                (composition-close . 0)
                (composition-open . 0)
                (cpp-define-intro c-lineup-cpp-define +)
                (cpp-macro . -1000)
                (cpp-macro-cont . +)
                (do-while-closure . 0)
                (extern-lang-close . 0)
                (extern-lang-open . 0)
                (friend . 0)
                (func-decl-cont . +)
                (inclass . +)
                (incomposition . +)
                (inexpr-class . +)
                (inexpr-statement . +)
                (inextern-lang . +)
                (inher-cont . c-lineup-multi-inher)
                (inher-intro . +)
                (inlambda . 0)
                (inline-close . 0)
                (inline-open . +)
                (inmodule . +)
                (innamespace . +)
                (knr-argdecl . 0)
                (knr-argdecl-intro . 0)
                (label . 0)
                (lambda-intro-cont . +)
                (module-close . 0)
                (module-open . 0)
                (namespace-close . 0)
                (namespace-open . 0)
                (objc-method-args-cont . c-lineup-ObjC-method-args)
                (objc-method-call-cont c-lineup-ObjC-method-call-colons c-lineup-ObjC-method-call +)
                (objc-method-intro .
                                   [0])
                (statement-case-open . 0)
                (stream-op . c-lineup-streamop)
                (string . -1000)
                (substatement . +)
                (substatement-label . 0)
                (template-args-cont c-lineup-template-args +)
                (topmost-intro-cont . c-lineup-topmost-intro-cont))))
```
I believe I can just add the change set to `c-offsets-alist` to `.dir-locals.el`. Anyway, I'll try to rebase the substantive part of this tomorrow, but I'm happy to hold `.dir-locals.el` for another PR if it seems like it could use more thought.

Reply:
> Do you actually "set your tab size to 4 characters" as the [Unreal guide](https://docs.unrealengine.com/5.1/en-US/epic-cplusplus-coding-standard-for-unreal-engine/#tabsandindenting) says?

Yes.  Tabs are to be 4 characters wide.

> but it seems wrong, and likewise GitHub displays, say, these lines: 

Yeah github does the wrong thing here.  I don't know that there's any way around it, but I don't think what github does matters all that much.  Here's what that same excerpt looks like when I work on it:

![image](https://user-images.githubusercontent.com/768449/221378448-afd0f730-2a64-4f9c-917f-1a6e7d0b5aff.png)

Yesterday I went through all of the style guides that correspond to the named formatting entries provided by emacs, and none of them match the relevant parts of the Unreal style guide.  For example, [the Linux style guide](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces) dictates brace placement rules that are incompatible (among other problems).

> I believe I can just add the change set to c-offsets-alist to .dir-locals.el. Anyway, I'll try to rebase the substantive part of this tomorrow, but I'm happy to hold .dir-locals.el for another PR if it seems like it could use more thought.

I think this is a reasonable conclusion.  From your `"STYLE NAME HERE"` excerpt above, it looks like emacs does provide you with the tools to do the correct thing here.  The style guide is significant, and I really do not want to have a file that says the project uses one that is not actually the correct style guide.

Reply:
> Yeah github does the wrong thing here. I don't know that there's any way around it, but I don't think what github does matters all that much. Here's what that same excerpt looks like when I work on it:

Yep, that definitely means it should be `4`. I'll change that, and I'll see what I can figure out about how github etc. are displaying it.

> > I believe I can just add the change set to c-offsets-alist to .dir-locals.el. Anyway, I'll try to rebase the substantive part of this tomorrow, but I'm happy to hold .dir-locals.el for another PR if it seems like it could use more thought.
> 
> I think this is a reasonable conclusion. From your `"STYLE NAME HERE"` excerpt above, it looks like emacs does provide you with the tools to do the correct thing here. The style guide is significant, and I really do not want to have a file that says the project uses one that is not actually the correct style guide.

Yes, committing a known-wrong configuration would definitely be bad! I'll update `.dir-locals.el` and try it out, but I'll move it to a separate PR, and maybe I can find someone who's familiar with the nuances of C-family indentation in Emacs to take a look at it.

Reply:
Moved to https://github.com/Aeva/tangerine/pull/11